src/cmd/pax/bax.1
.\"
.\" Glenn Fowler
.\" AT&T Bell Laboratories
.\"
.\" @(#)bax.1 (gsf@research.att.com) 05/09/95
.\"
.fp 5 CW
.TH BAX 1
.SH NAME
bax \- tw + pax delta backup
.SH SYNOPSIS
.B bax
.I file
.I base
[
.I "dir ..."
]
.SH DESCRIPTION
.I bax
generates backup tapes using
.I tw
and
.IR pax .
.I file
is the name of the backup archive to be generated.
If
.I base
is
.L \-
then a full backup is generated in
.IR file .
Otherwise
.I base
names a previous full backup against which a delta backup archive is
generated in 
.IR file .
The backup starts at the directories
.IR "dir ..." ,
.L /
by default.
File system boundaries are not crossed;  multiple file systems
must be specified as separate
.I dir
arguments.
.P
Variables defined in the shell script
.L ../lib/bax/init
located on
.L $PATH
customize the files and directories to be backed up.
The variables and their default values are:
.TP
\fBarch_name\f5='arch'\fR
.TP
\fBarch_keep\f5='bin|fun|lib'\fR
Shell patterns that match the architecture specific root directories and files.
The architecture specific files that match
\f5*/${\fIarch_name\f5}/*/${\fIarch_keep\f5}\fR are retained;
all other architecture specific files are skipped.
.TP
\fBbig_size\f5=5000000\fR
.TP
\fBbig_magic_keep\f5='mail'\fR
Files larger than \f5${\fIbig_size\f5}\fR must have a magic file type
that matches \f5${\fIbig_magic_keep\f5}\fR.
.TP
\fBname_skip\f5'?(*.)(old|save|tmp)?(.*)|.*-cache'\fR
File base names that match \f5${\fIname_skip\f5}\fR are skipped.
.TP
\fBpath_skip\f5='/dev|/usr/local/ast/ship'\fR
File path names that match \f5${\fIpath_skip\f5}\fR are skipped.
.TP
\fBreg_magic_skip\f5='core|cql * index|history|TeX dvi'\fR
Regular files with a magic file type that matches \f5${\fIreg_magic_skip\f5}\fR
are skipped.
.TP
\fBsrc_name\f5='src'\fR
.TP
\fBsrc_magic_skip\f5='executable|object|(archive|shared) library'\fR
Files under \f5*/${\fIsrc_name\f5}/*\fR with a magic file type that
matches \f5${\fIreg_magic_skip\f5}\fR are skipped.
.SH EXAMPLES
To generate a full backup on \f5/dev/st0\fP:
.EX
bax /dev/st0 -
.EE
To generate a delta backup in \f5/tmp/delta.bax\fP for the full backup above:
.EX
bax /tmp/delta.bax /dev/st0
.EE
or if you have two tape drives:
.EX
mt /dev/st1 eod # to append the next delta
bax /dev/st1 /dev/st0
.EE
.SH "SEE ALSO"
file(1), pax(1), tw(1), vdelta(3)


src/cmd/pax/bax.sh
########################################################################
#                                                                      #
#               This software is part of the ast package               #
#          Copyright (c) 1987-2011 AT&T Intellectual Property          #
#                      and is licensed under the                       #
#                 Eclipse Public License, Version 1.0                  #
#                    by AT&T Intellectual Property                     #
#                                                                      #
#                A copy of the License is available at                 #
#          http://www.eclipse.org/org/documents/epl-v10.html           #
#         (with md5 checksum b35adb5213ca9657e911e9befb180842)         #
#                                                                      #
#              Information and Software Systems Research               #
#                            AT&T Research                             #
#                           Florham Park NJ                            #
#                                                                      #
#                 Glenn Fowler <gsf@research.att.com>                  #
#                                                                      #
########################################################################
: tw + pax for backup
#
# @(#)bax (AT&T Bell Laboratories) 05/09/95
#
# ../lib/bax/init on $PATH to customize
#

command=bax
customize=../lib/${command}/init

#
# no stinking environment
#

root= arch_name= arch_keep= big_size= big_magic_keep=
name_skip= path_skip= reg_magic_skip= src_name= src_magic_skip=

ifs=${IFS-'
	 '}
IFS=:
path=$PATH
IFS=$ifs
for dir in $path
do	if	test -f $dir/$customize
	then	. $dir/$customize
		break
	fi
done

#
# default dirs
#

root=${root:-'/'}

#
# keep these architecture specific files: */${arch_name}/*/${arch_keep}
#

arch_name=${arch_name:-'arch'}
arch_keep=${arch_keep:-'bin|fun|lib'}

#
# files bigger than ${big_size} must have magic that matches ${big_magic_keep}
#

big_size=${big_size:-5000000}
big_magic_keep=${big_magic_keep:-'mail'}

#
# ${name_skip} are base names to be skipped
# ${path_skip} are path prefixes to be skipped
#

name_skip=${name_skip:-'?(*.)(old|save|tmp)?(.*)|.*-cache'}
path_skip=${path_skip:-'/dev|/usr/local/ast/ship'}

#
# regular files with magic that matches ${reg_magic_skip} are skipped
#

reg_magic_skip=${reg_magic_skip:-'core|cql * index|history|TeX dvi'}

#
# files under */${src_name}/* are skipped if magic matches ${src_magic_skip}
#

src_name=${src_name:-'src'}
src_magic_skip=${src_magic_skip:-'executable|object|(archive|shared) library'}

case $# in
[01])	echo "Usage: $command delta base [dir ...]" >&2
	exit 2
	;;
esac
delta=$1
shift
base=$2
shift
case $# in
0)	set $root ;;
esac
dirs=
while	:
do	case $# in
	0)	break ;;
	esac
	dirs="$dirs -d $1"
	shift
done

tw $dirs -mH -e "

begin:	int	arch_bit = 1;
	int	src_bit = 2;
	char*	m;

sort:	name;

select:	if (name == '${name_skip}')
		status = SKIP;
	else if (type == REG)
	{
		m = magic;
		if (size > ${big_size} && m != '*@(${big_magic_keep})*' ||
		    m == '*@(${reg_magic_skip})*' ||
		    (parent.local & src_bit) && m == '*@(${src_magic_skip})*')
			status = SKIP;
	}
	else if (type == DIR)
	{
		local = (parent.local & src_bit);
		if (dev != parent.dev || path == '${path_skip}' || (parent.parent.local & arch_bit) && name != '${arch_keep}')
			status = SKIP;
		else if (name == '${arch_name}')
			local |= arch_bit;
		else if (name == '${src_name}')
			local |= src_bit;
	}
	return 1;
" | pax -o "write ordered physical verbose file=$delta base=$base"
